sphinx/themes/basic/static/searchtools.js

```diff
@@ -219,7 +219,8 @@ const Search = {
     (document.body.appendChild(document.createElement("script")).src = url),
 
   setIndex: (index) => {
-    Search._index = index;
+    const reviver = (k, v) => (typeof v === "object" && v !== null) ? Object.freeze(v) : v;
+    Search._index = JSON.parse(index, reviver);
```

Comment:
I think @wlach is right that immutability is orthoganal here. I think 'simply' using `JSON.parse` rather than an object literal would solve the `__proto__` case. Would we then need to keep the `hasOwnProperty` checks?

An alternative would be to use a reviver function that constructs `Map` objects/serialise the index with `Map` literals, and use `.get()`, but that may also be expensive.

```suggestion
    Search._index = JSON.parse(index);
```

Reply:
> I think @wlach is right that immutability is orthoganal here. I think 'simply' using `JSON.parse` rather than an object literal would solve the `__proto__` case.

Ok, yep - I can revert 443fd213c75d5ba7eb9b04869ccad8ca655f2393 to remove the use of `Object.freeze` / the reviver function.
 
>  Would we then need to keep the `hasOwnProperty` checks?

We should keep those, I think.  Without them, queries for `__proto__` may continue to inadvertently retrieve object prototypes instead of getting an `undefined` result.

> An alternative would be to use a reviver function that constructs `Map` objects/serialise the index with `Map` literals, and use `.get()`, but that may also be expensive.

What do you think about constructing those `Map` objects directly in the `searchindex.js` file format, instead of using `JSON.parse` and a reviver?

Reply:
> What do you think about constructing those Map objects directly in the searchindex.js file format, instead of using JSON.parse and a reviver?

IMO this is definitely worth a shot, I suspect it'll be quite fast.

Reply:
I'm giving it a try - it doesn't seem too tricky; the Python serialization of the search index becomes a little more complex.  One challenge that I have yet to work out is how the Python unit tests should parse/load the resulting index (ref: https://github.com/sphinx-doc/sphinx/blob/15b46dc81e56ac1fd352da808f7aadd3e825a101/tests/test_search.py#L70).